Q1DWeighted to respect bin masking and gravity

gvardany opened the pull request:
**Description of work.**

This PR extends the `Q1DWeighted` algorithm to discard the masked bins in the input, if present, and optionally take the analytical gravity drop into account when calculating the Q. Besides, it improves the performance significantly for TOF mode due to reorganisation of the code.

**To test:**

0. Code review & tests pass.

1.
```
# data is in UnitTest data
ws = Load('ILL_D33_LTOF_red.nxs')
ws = MaskBinsIf(InputWorkspace=ws, Criterion='x < 1 || x > 10')
q1dw = Q1DWeighted(InputWorkspace=ws, OutputBinning='0.005,-0.1,1.', NumberOfWedges=0)
q1dwg = Q1DWeighted(InputWorkspace=ws, OutputBinning='0.005,-0.1,1.', NumberOfWedges=0, AccountForGravity=True)
```
Overlay the 2 outputs on log-log scale, zoom in to the low Q part. There should be a small difference there, and only there.

2. Time the performance test. Copy it, and time it also on this [commit](https://github.com/mantidproject/mantid/pull/24699/commits/8fc266ff479d68a975437408f4df296ba2087aac). 

<!-- Instructions for testing. -->

Fixes #24686, #24701. <!-- and fix #xxxx or close #xxxx xor resolves #xxxx -->

---
#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

SimonHeybrock commented on Framework/Algorithms/src/Q1DWeighted.cpp:
```diff
@@ -228,10 +244,22 @@ void Q1DWeighted::exec() {
+          const double drop = PhysicalConstants::g * tof * tof / 2.;
+          V3D correction(0., drop, 0.);
+          pos += correction;
+        }
```
What is the reasons for not using https://github.com/mantidproject/mantid/blob/master/Framework/Algorithms/inc/MantidAlgorithms/GravitySANSHelper.h here?

gvardany replied:
Good point, I will have a look. 👍

gvardany commented on Framework/Algorithms/src/Q1DWeighted.cpp:
```diff
@@ -207,14 +214,23 @@ void Q1DWeighted::exec() {
         continue;
 
+      // Skip the masked bins
+      if (inputWS->hasMaskedBins(i)) {
```
You are right that it is very slow and redundant for the masked bins. I was originally planning to reorganise the code in a separate issue (see #24701), which was promising orders of magnitude of speed up. Until this morning that I realised that there is a reason why it loops first over lambda. So I think I will actually need to implement some caching of geometry and bin masking information, since they do not depend on the wavelength bin. Marking In Progress meanwhile.

gvardany commented:
I have refactored the code in this PR at the end. This should be an order of magnitude faster than before. Note that I had to update system test references, since the refactoring also fixed a small bug:
[here](https://github.com/mantidproject/mantid/blob/master/Framework/Algorithms/src/Q1DWeighted.cpp#L235) the Q bin index is initialised with 0, but [here](https://github.com/mantidproject/mantid/blob/master/Framework/Algorithms/src/Q1DWeighted.cpp#L256) 0 also passes; that means that if the Q was out of range, the bin was counted in the first Q bin still. I will post some to test instructions shortly.

gvardany commented:
@SimonHeybrock This is ready for a second review.

SimonHeybrock commented on Framework/Algorithms/src/Q1DWeighted.cpp:
```diff
@@ -154,199 +139,228 @@ void Q1DWeighted::exec() {
+  // the same dimensions for the normalisation weights
+  std::vector<std::vector<std::vector<double>>> normalisation(
+      nWedges + 1,
+      std::vector<std::vector<double>>(nLambda, std::vector<double>(nQ, 0.0)));
```
Can you just copy `errors` and `normalisation` from `intensities`, to avoid the repeated lengthy init code?

SimonHeybrock commented on Framework/Algorithms/src/Q1DWeighted.cpp:
```diff
@@ -154,199 +139,228 @@ void Q1DWeighted::exec() {
-        }
+      V3D correction;
+      if (correctGravity) {
+        correction.setY(gravityHelper.gravitationalDrop(wavelength));
```
Assumes that `Y` is the up direction? Should this use some information from the instrument instead?

SimonHeybrock commented:
No answer from the other SANS developers yet? I'll just approve this for now, maybe the gatekeeper can find a reviewer, or we leave it to the upcoming beta testing?

gvardany commented:
@SimonHeybrock Yes I think we leave for the gatekeeper, had to resolve a conflict in release notes, can you approve again?

martyngigg commented:
ISIS uses `Q1D`. HFIR and EQSANS use `Q1DWeighted` and they have pretty decent system test coverage. Those haven't required updating so I'm happy with merging this.
